### PR TITLE
Initial PR - Create transport

### PR DESCRIPTION
**Pull request by RLRabinowitz**
This PR creates the transport. It's very similar to the implementation of https://github.com/lazywithclass/winston-cloudwatch, with some adjustments for DynamoDB:

Index.js:
- [x] Throw exception when region doesn't support DynamoDB (partially taken from https://github.com/jwchang0206/winston-dynamodb)
- [x] Remove `logGroupName` and `retentionInDays`, which is CW specific, and add `tableName` instead
- [x] Change AWS config format to match that of DynamoDB client 
- [x] Rename every CW to DynamoDB

util.js:
- [x] Rename `WINSTON_CLOUDWATCH_DEBUG` to `WINSTON_DYNAMODB_DEBUG`

dynamodb-integration.js:
- [x] Remove all unnecessary logic from cloudwatch-integration.js - Managing tokens and related errors, not being able to send multiple batch requests simultaniously...
- [x] Keep size limit logic + message truncation
- [x] Do not allow more than 25 items in a batch
- [x] Creates a batch request. The `id` is the `streamName`, `timestamp` is the timestamp of log addition, `message` is the log message

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+var util = require('util'),
+    winston = require('winston'),
+    AWS = require('aws-sdk'),
+    dynamodbIntegration = require('./lib/dynamodb-integration'),
+    isEmpty = require('lodash.isempty'),
+    assign = require('lodash.assign'),
+    isError = require('lodash.iserror'),
+    stringify = require('./lib/utils').stringify,
+    debug = require('./lib/utils').debug,
+    defaultFlushTimeoutMs = 10000;
```
Comment (avnerenv0) [inline index.js]: put a `const` in each line?

File: index.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    debug('log (called by winston)', info);
+
+    if (!isEmpty(info.message) || isError(info.message)) {
+        this.add(info);
+    }
+
+    if (!/^uncaughtException: /.test(info.message)) {
+        // do not wait, just return right away
+        return callback(null, true);
+    }
```
Comment (avnerenv0) [inline index.js:84]: this one is confusing - maybe reverse the condition

File: index.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    // as Winston is about to end the process
+    clearInterval(this.intervalId);
+    this.intervalId = null;
+    this.submit(callback);
+};
+
+WinstonDynamoDB.prototype.add = function(log) {
+    debug('add log to queue', log);
+
+    var self = this;
```
Comment (avnerenv0) [inline index.js]: maybe use arrow functions and get rid of this shameful piece of javascript history

File: index.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        this.dynamoDB,
+        this.tableName,
+        streamName,
+        this.logEvents,
+        this.options,
+        callback
+    );
+};
+
+WinstonDynamoDB.prototype.kthxbye = function(callback) {
```
Comment (avnerenv0) [inline index.js:136]: `kthxbye` is unique to that transport right? 
Doesn't winston have something more 'built-in' ?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    clearInterval(this.intervalId);
+    this.intervalId = null;
+    this.flushTimeout = this.flushTimeout || (Date.now() + defaultFlushTimeoutMs);
+
+    this.submit((function(error) {
+        if (error) return callback(error);
+        if (isEmpty(this.logEvents)) return callback();
+        if (Date.now() > this.flushTimeout) return callback(new Error('Timeout reached while waiting for logs to submit'));
+        else setTimeout(this.kthxbye.bind(this, callback), 0);
+    }).bind(this));
```
Comment (avnerenv0) [inline index.js:146]: you are calling `this.submit` I don't think you need this `bind(this);`

File: index.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+};
+
+winston.transports.CloudWatch = WinstonDynamoDB;
```
Comment (avnerenv0) [inline index.js]:
```suggestion
winston.transports.DynamoDB = WinstonDynamoDB;
```

File: index.js [hunk 7]
```diff
@@ -0,0 +1,153 @@
+'use strict';
```
Comment (avnerenv0) [inline index.js:1]: I don't think `use strict` is needed anymore anywhere

File: lib/dynamodb-integration.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
+var LIMITS = {
```
Comment (avnerenv0) [inline lib/dynamodb-integration.js]: prefer `const`

File: jest.config.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+/*
```
Comment (avnerenv0) [inline jest.config.js]: why is this needed if there aren't any tests?

File: lib/dynamodb-integration.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        if (bytes + evSize > LIMITS.MAX_BATCH_SIZE_BYTES) break;
+        bytes += evSize;
+        entryIndex++;
+    }
+
+    const eventsToSend = logEvents.splice(0, entryIndex);
+    const payload = {
+        RequestItems: {
+            [tableName]: eventsToSend.map(event =>
+                ({
```
Comment (avnerenv0) [inline lib/dynamodb-integration.js:41]: The `streamName` will be the partition key and the `timestamp` sort key?

File: lib/dynamodb-integration.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            evSize = LIMITS.MAX_EVENT_MSG_SIZE_BYTES;
+            ev.message = ev.message.substring(0, evSize);
+            const msgTooBigErr = new Error('Message Truncated because it exceeds the DynamoDB size limit');
+            msgTooBigErr.logEvent = ev;
+            cb(msgTooBigErr);
+        }
+        if (bytes + evSize > LIMITS.MAX_BATCH_SIZE_BYTES) break;
+        bytes += evSize;
+        entryIndex++;
+    }
```
Comment (avnerenv0) [inline lib/dynamodb-integration.js:32]: I would put all this in a different function

File: lib/dynamodb-integration.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    while (entryIndex < Math.min(logEvents.length, LIMITS.MAX_BATCH_ITEM_NUM)) {
+        var ev = logEvents[entryIndex];
+        // unit tests pass null elements
+        var evSize = ev ? Buffer.byteLength(ev.message, 'utf8') + BASE_EVENT_SIZE_BYTES : 0;
+        if(evSize > LIMITS.MAX_EVENT_MSG_SIZE_BYTES) {
+            evSize = LIMITS.MAX_EVENT_MSG_SIZE_BYTES;
+            ev.message = ev.message.substring(0, evSize);
+            const msgTooBigErr = new Error('Message Truncated because it exceeds the DynamoDB size limit');
+            msgTooBigErr.logEvent = ev;
+            cb(msgTooBigErr);
```
Comment (avnerenv0) [inline lib/dynamodb-integration.js:27]: this calls the callback but continues the run ? doesn't seem right, maybe missing a `return` ?
Also maybe instead we can split the large message to several smaller messages ?

**Review by avnerenv0 — APPROVED** [review]
@RLRabinowitz This looks good enough to use, but I reviewed it like 'new code', feel free to apply comments however you like

**Comment by avnerenv0** [comment]
I would also add a Dynamo DB Table definition file

File: index.js (hunk 4 above)
Comment (RLRabinowitz) [reply index.js:136]: Yep it's unique to the transport

File: lib/dynamodb-integration.js (hunk 10 above)
Comment (RLRabinowitz) [reply lib/dynamodb-integration.js:41]: In this PR, yes. I'll change the `timestamp` to be a `sequenceNumber` in a next PR, because of [this issue](https://github.com/env0/winston-aws-dynamodb/issues/3)

File: lib/dynamodb-integration.js (hunk 12 above)
Comment (RLRabinowitz) [reply lib/dynamodb-integration.js:27]: Thought about it, but since this is the existent logic in CW transport, I've decided not to touch this.
Also, the size limit for the truncation is bigger on DynamoDB (400KB) than in CW (256KB), so I wouldn't worry about it for our use case

**Comment by RLRabinowitz** [comment]
> I would also add a Dynamo DB Table definition file

Add this to the [missing README issue](https://github.com/env0/winston-aws-dynamodb/issues/2)

File: jest.config.js (hunk 9 above)
Comment (RLRabinowitz) [reply jest.config.js]: Right. This isn't need yet. I can remove it and add it later when necessary

File: index.js (hunk 7 above)
Comment (RLRabinowitz) [reply index.js:1]: I'm not too familiar with this, but removing `'use strict'` seems to allow stuff like assigning undeclared variables (tested this locally now)

File: index.js (hunk 5 above)
Comment (RLRabinowitz) [reply index.js:146]: Yeah that makes sense. However I don't think I'll change that (yet, at least) since I didn't test this method yet

File: index.js (hunk 3 above)
Comment (RLRabinowitz) [reply index.js]: Interesting, read online and didn't know about the whole `this` scope thing that is different between `function`s and arrow functions.
TIL

File: index.js (hunk 2 above)
Comment (RLRabinowitz) [reply index.js:84]: Since I'm not too sure on this code here (Specifically what the `callback(null, true) means`. It seems that the callback is called immediately with these params on `logger.info`, but is called again with other params when the batch is sent), I'll probably skip this...

Comment (avnerenv0) [reply index.js:84]: callbacks usually follow the pattern of `func(error, result)` and in this case `cb(null, true)` means everything worked. This is the 'happy flow' and will happen for EVERY log call except when the message includes `uncaughtException`. 

It's confusing because the 'normal' case is inside the if and short circuits this function.

If the error message does include `uncaughtException` then we clear everything and submit the logs and get ready for a 'crash' exit of the process.

File: index.js (hunk 1 above)
Comment (RLRabinowitz) [reply index.js]: 7e0822261e5961881f3329083700be629451d6ac

File: index.js (hunk 6 above)
Comment (RLRabinowitz) [reply index.js]: 0a1675e19e9fd392f110becec526d2ae56cd6b94

File: index.js (hunk 3 above)
Comment (RLRabinowitz) [reply index.js]: 230fc51df2dc879ebc9f55f94973d726ed0361fd

File: lib/dynamodb-integration.js (hunk 8 above)
Comment (RLRabinowitz) [reply lib/dynamodb-integration.js]: 7e0822261e5961881f3329083700be629451d6ac

File: lib/dynamodb-integration.js (hunk 11 above)
Comment (RLRabinowitz) [reply lib/dynamodb-integration.js:32]: e1364211f2fedce820e21d9ca430a4834c6a925c

File: jest.config.js (hunk 9 above)
Comment (RLRabinowitz) [reply jest.config.js]: 70d22ef22b38d6dd2ffa7537597b09fde0ae184e

**Comment by RLRabinowitz** [comment]
[DynamoDB winston transport](https://trello.com/c/VHScBg03/3485-dynamodb-winston-transport)
